charms/kfp-api/src/charm.py

```diff
@@ -403,23 +409,147 @@ def _validate_sdi_interface(self, interfaces: dict, relation_name: str, default_
         # Remove once addressed in:
         # https://github.com/canonical/serialized-data-interface/issues/28
         if len(data_dict) == 0:
-            raise CheckFailedError(
-                f"Found incomplete/incorrect relation data for {relation_name}.",
+            raise ErrorWithStatus(
+                f"Found empty relation data for {relation_name}",
                 BlockedStatus,
             )
 
         return data_dict
 
+    def _get_mysql(self):
+        mysql = self.model.relations["mysql"]
+        if len(mysql) == 0:
+            raise ErrorWithStatus("Please add required relation mysql", BlockedStatus)
+        elif len(mysql) > 1:
+            raise ErrorWithStatus("Too many mysql relations", BlockedStatus)
+
+        try:
+            mysql = mysql[0]
+            unit = list(mysql.units)[0]
+            mysql = mysql.data[unit]
+        except Exception as e:
+            self.logger.error(
+                f"Encountered the following exception when parsing mysql relation: " f"{str(e)}"
+            )
+            raise ErrorWithStatus(
+                "Unexpected error when parsing mysql relation. See logs", BlockedStatus
+            )
 
-class CheckFailedError(Exception):
-    """Raise this exception if one of the checks in main fails."""
+        expected_attributes = ["database", "host", "root_password", "port"]
 
-    def __init__(self, msg, status_type=None):
-        super().__init__()
+        missing_attributes = [
+            attribute for attribute in expected_attributes if attribute not in mysql
+        ]
 
-        self.msg = msg
-        self.status_type = status_type
-        self.status = status_type(msg)
+        if len(missing_attributes) == len(expected_attributes):
+            raise ErrorWithStatus("Waiting for mysql relation data", WaitingStatus)
+        elif len(missing_attributes) > 0:
+            self.logger.error(
+                f"mysql relation data missing expected attributes '{missing_attributes}'"
+            )
+            raise ErrorWithStatus(
+                "Received incomplete data from mysql relation. See logs", BlockedStatus
+            )
+        return mysql
+
+    def _get_object_storage(self, interfaces):
+        """Retrieve object-storage relation data."""
+        relation_name = "object-storage"
+        object_storage_data = None
+        try:
+            object_storage_data = self._validate_sdi_interface(interfaces, relation_name)
+        except ErrorWithStatus as error:
+            raise error
```

Comment:
```suggestion
        object_storage_data = self._validate_sdi_interface(interfaces, relation_name)
```

Reply:
Why not have exception handling here?
I would like to have an error raised, because I don't want to proceed with container configuration if object interface validation failed. Container config relies on some data from object storage relation.

Reply:
Ok. Then I need to re-think this SDI interface validation.

Reply:
Remove exception handling.